### ISSUE TITLE
[OCCT>7.5.0] Various errors when try to create an SVG

### ISSUE THREAD

**Issue by theoryshaw** [IfcOpenShell/IfcOpenShell#4389]
Using [this](https://hub.openingdesign.com/OpeningDesign/Highland_Haven/src/commit/f641942869d20542fe79a6f42713c297dada05bd/Open/Models_and_CAD/BlenderBIM/Highland_Haven.ifc) file, when i try and create a SVG with following drawings, i get various errors.  Other drawings in the file work, but some don't... can't narrow down the settings that causing this. 

drawing: `SECTION - SOUTH - GARAGE`

```
Python: Traceback (most recent call last):
  File "C:\Users\Ryan Schultz\AppData\Roaming\Blender Foundation\Blender\3.6\scripts\addons\blenderbim\bim\module\drawing\operator.py", line 212, in invoke
    return self.execute(context)
  File "C:\Users\Ryan Schultz\AppData\Roaming\Blender Foundation\Blender\3.6\scripts\addons\blenderbim\bim\module\drawing\operator.py", line 237, in execute
    self.metadata = tool.Drawing.get_drawing_metadata(self.camera_element)
  File "C:\Users\Ryan Schultz\AppData\Roaming\Blender Foundation\Blender\3.6\scripts\addons\blenderbim\tool\drawing.py", line 1507, in get_drawing_metadata
    for v in ifcopenshell.util.element.get_psets(drawing)["EPset_Drawing"].get("Metadata", "").split(",")
AttributeError: 'NoneType' object has no attribute 'split'

```

drawing: `SECTION - SOUTH - KITCHEN`

```
Python: Traceback (most recent call last):
  File "C:\Users\Ryan Schultz\AppData\Roaming\Blender Foundation\Blender\3.6\scripts\addons\blenderbim\bim\module\drawing\operator.py", line 212, in invoke
    return self.execute(context)
  File "C:\Users\Ryan Schultz\AppData\Roaming\Blender Foundation\Blender\3.6\scripts\addons\blenderbim\bim\module\drawing\operator.py", line 263, in execute
    linework_svg = self.generate_linework(context)
  File "C:\Users\Ryan Schultz\AppData\Roaming\Blender Foundation\Blender\3.6\scripts\addons\blenderbim\bim\module\drawing\operator.py", line 526, in generate_linework
    self.serialize_contexts_elements(ifc, tree, contexts, "body", drawing_elements, target_view)
  File "C:\Users\Ryan Schultz\AppData\Roaming\Blender Foundation\Blender\3.6\scripts\addons\blenderbim\bim\module\drawing\operator.py", line 471, in serialize_contexts_elements
    self.serialiser.write(elem)
  File "C:\Users\Ryan Schultz\AppData\Roaming\Blender Foundation\Blender\3.6\scripts\addons\blenderbim\libs\site\packages\ifcopenshell\ifcopenshell_wrapper.py", line 1512, in write
    return _ifcopenshell_wrapper.SvgSerializer_write(self, *args)
RuntimeError: Invalid cast
```

drawing: `WEST SECTION`

```
Python: Traceback (most recent call last):
  File "C:\Users\Ryan Schultz\AppData\Roaming\Blender Foundation\Blender\3.6\scripts\addons\blenderbim\bim\module\drawing\operator.py", line 212, in invoke
    return self.execute(context)
  File "C:\Users\Ryan Schultz\AppData\Roaming\Blender Foundation\Blender\3.6\scripts\addons\blenderbim\bim\module\drawing\operator.py", line 263, in execute
    linework_svg = self.generate_linework(context)
  File "C:\Users\Ryan Schultz\AppData\Roaming\Blender Foundation\Blender\3.6\scripts\addons\blenderbim\bim\module\drawing\operator.py", line 526, in generate_linework
    self.serialize_contexts_elements(ifc, tree, contexts, "body", drawing_elements, target_view)
  File "C:\Users\Ryan Schultz\AppData\Roaming\Blender Foundation\Blender\3.6\scripts\addons\blenderbim\bim\module\drawing\operator.py", line 471, in serialize_contexts_elements
    self.serialiser.write(elem)
  File "C:\Users\Ryan Schultz\AppData\Roaming\Blender Foundation\Blender\3.6\scripts\addons\blenderbim\libs\site\packages\ifcopenshell\ifcopenshell_wrapper.py", line 1512, in write
    return _ifcopenshell_wrapper.SvgSerializer_write(self, *args)
RuntimeError: Token $ at offset 24,443,121 invalid instance name
```

**Comment by Moult** [comment]
The third error is in C++, where it falls over when a pset contains a null value (which is valid in IFC, but I guess unexpected). I've committed a fix but ping @aothms if I messed it up.

I did encounter another issue which I was unable to dig deeper for `SECTION - SOUTH - GARAGE`:

```
Error: Python: Traceback (most recent call last):
  File "/home/dion/.config/blender/4.0/scripts/addons/blenderbim/bim/module/drawing/operator.py", line 263, in execute
    linework_svg = self.generate_linework(context)
  File "/home/dion/.config/blender/4.0/scripts/addons/blenderbim/bim/module/drawing/operator.py", line 540, in generate_linework
    self.serialiser.finalize()
  File "/home/dion/.config/blender/4.0/scripts/addons/blenderbim/libs/site/packages/ifcopenshell/ifcopenshell_wrapper.py", line 1153, in finalize
    return _ifcopenshell_wrapper.SvgSerializer_finalize(self)
RuntimeError: An unknown error occurred
```

If I exclude this element the error goes away - it's purely edges (section cut is annotated) - so I guess the cut only results in ... verts? and that confuses the HLR?:

![2024-03-04-174847_709x695_scrot](https://github.com/IfcOpenShell/IfcOpenShell/assets/88302/c23bf36e-be59-48e0-8dea-d1f73d71596d)

**Comment by Moult** [comment]
@aothms my guess is that this scenario can't be reliably detected before feeding into HLRBRep_*Algo, so in SvgSerializer.h in `std::list<std::pair<const IfcUtil::IfcBaseEntity*, TopoDS_Shape>> build()` maybe instead of directly `return boost::apply_visitor(vis, engine_);` then do something like:

```c++
// somehow try catch failed items here
auto result = boost::apply_visitor(vis, engine_); // This is where the Runtime Error is triggered
return result;

```

**Comment by aothms** [comment]
The null fix looks great, thank you. The other suggestion I don't really understand.

The SVG code is a bit hard to follow due to the all the typing indirections:

```C++
typedef prefiltered_hlr hlr_t

|
v

typedef boost::variant<
	boost::blank,
	Handle(HLRBRep_Algo),
	Handle(HLRBRep_PolyAlgo)
> hlr_brep_or_poly_t;

|
v

class hlr_calc;
```

So my guess is that the error is actually in `algo->Update();` or `algo->Hide();`. Which are global for each shape, which means we have to detect this before feeding into HLRBRep_*Algo.

But we can do that. At least in this case, it's really rather quick to assert that the wire is orthogonal to the view direction.

Is there a way that I can reproduce this from the command line so that I can attach a debugger? Maybe, can we somehow automate this, that users get to copy paste a script to reproduce drawing relating errors without blenderbim? I don't want to sound lazy, but it would really help me, because it's hard for me to keep up with the amazing pace of the addon development.

**Comment by Moult** [comment]
In short the error is triggered in that boost apply visitor line (courtesy of liberal print statements :)

I'll produce a stand alone script for you to recreate :) searching for svg now shows there's 59 open issues and I'm sure there are some duplicates and recurring issues. I'll work through them methodically and isolate them the best I can for you.

@theoryshaw for a quick fix right now you can edit the drawing pset to remove any nulls, and then hide any 2d stuff with traits similar to the screenshot.

Dion Moult

Sent from Proton Mail mobile

-------- Original Message --------

**Comment by Moult** [comment]
@aothms here's a standalone script to recreate this bug:

Run it like this (first guid is the guid of the drawing element, second is the troublesome element):

```
python debug.py Highland_Haven.ifc 2fUzMgD8HBUxSnHecfYzsZ 3KoCvuOo1C1PWj3aOf1esq
```

```python
import sys
import ifcopenshell
import ifcopenshell.geom
import multiprocessing
from typing import NamedTuple, List

# python standalone_drawer.py model.ifc guid_of_drawing guid_of_bad_element output.svg


class LineworkContexts(NamedTuple):
    body: List[List[int]]
    annotation: List[List[int]]


class Drawer:
    def execute(self):
        ifc = ifcopenshell.open(sys.argv[1])
        self.camera_element = ifc.by_guid(sys.argv[2])
        # Don't use draw.main() just whilst we're prototyping and experimenting
        # Get all representation contexts to see what we're dealing with.
        target_view = ifcopenshell.util.element.get_psets(self.camera_element)["EPset_Drawing"]["TargetView"]
        contexts = self.get_linework_contexts(ifc, target_view)
        drawing_elements = {ifc.by_guid(sys.argv[3])}

        self.setup_serialiser(ifc, target_view)

        tree = ifcopenshell.geom.tree()
        tree.enable_face_styles(True)

        self.serialize_contexts_elements(ifc, tree, contexts, "body", drawing_elements, target_view)
        self.serialize_contexts_elements(ifc, tree, contexts, "annotation", drawing_elements, target_view)

        if self.camera_element not in drawing_elements:
            # The camera must always be included, regardless of any include/exclude filters.
            geom_settings = ifcopenshell.geom.settings(DISABLE_TRIANGULATION=True, STRICT_TOLERANCE=True)
            it = ifcopenshell.geom.iterator(geom_settings, ifc, include=[self.camera_element])
            for elem in it:
                self.serialiser.write(elem)

        self.serialiser.finalize()
        results = self.svg_buffer.get_value()
        print("results", results)

        with open(sys.argv[4], "w") as svg:
            svg.write(results)

    def get_linework_contexts(self, ifc, target_view) -> LineworkContexts:
        plan_body_target_contexts = []
        plan_body_model_contexts = []
        model_body_target_contexts = []
        model_body_model_contexts = []

        plan_annotation_target_contexts = []
        plan_annotation_model_contexts = []
        model_annotation_target_contexts = []
        model_annotation_model_contexts = []

        for rep_context in ifc.by_type("IfcGeometricRepresentationContext"):
            if rep_context.is_a("IfcGeometricRepresentationSubContext"):
                if rep_context.ContextType == "Plan":
                    if rep_context.ContextIdentifier in ["Body", "Facetation"]:
                        if rep_context.TargetView == target_view:
                            plan_body_target_contexts.append(rep_context.id())
                        elif rep_context.TargetView == "MODEL_VIEW":
                            plan_body_model_contexts.append(rep_context.id())
                    elif rep_context.ContextIdentifier == "Annotation":
                        if rep_context.TargetView == target_view:
                            plan_annotation_target_contexts.append(rep_context.id())
                        elif rep_context.TargetView == "MODEL_VIEW":
                            plan_annotation_model_contexts.append(rep_context.id())
                elif rep_context.ContextType == "Model":
                    if rep_context.ContextIdentifier in ["Body", "Facetation"]:
                        if rep_context.TargetView == target_view:
                            model_body_target_contexts.append(rep_context.id())
                        elif rep_context.TargetView == "MODEL_VIEW":
                            model_body_model_contexts.append(rep_context.id())
                    elif rep_context.ContextIdentifier == "Annotation":
                        if rep_context.TargetView == target_view:
                            model_annotation_target_contexts.append(rep_context.id())
                        elif rep_context.TargetView == "MODEL_VIEW":
                            model_annotation_model_contexts.append(rep_context.id())
            elif rep_context.ContextType == "Model":
                # You should never purely assign to a "Model" context, but
                # if you do, this is what we assume your intention is.
                model_body_model_contexts.append(rep_context.id())
                continue

        body_contexts = (
            [
                plan_body_target_contexts,
                plan_body_model_contexts,
                model_body_target_contexts,
                model_body_model_contexts,
            ]
            if target_view in ["PLAN_VIEW", "REFLECTED_PLAN_VIEW"]
            else [
                model_body_target_contexts,
                model_body_model_contexts,
            ]
        )

        annotation_contexts = (
            [
                plan_annotation_target_contexts,
                plan_annotation_model_contexts,
                model_annotation_target_contexts,
                model_annotation_model_contexts,
            ]
            if target_view in ["PLAN_VIEW", "REFLECTED_PLAN_VIEW"]
            else [
                model_annotation_target_contexts,
                model_annotation_model_contexts,
            ]
        )

        return LineworkContexts(body_contexts, annotation_contexts)

    def serialize_contexts_elements(
        self, ifc, tree: ifcopenshell.geom.tree, contexts: LineworkContexts, context_type, drawing_elements, target_view
    ):
        drawing_elements = drawing_elements.copy()
        contexts = getattr(contexts, context_type)
        for context in contexts:
            if not context or not drawing_elements:
                continue
            geom_settings = ifcopenshell.geom.settings(
                DISABLE_TRIANGULATION=True, STRICT_TOLERANCE=True, INCLUDE_CURVES=True
            )
            if ifc.by_id(context[0]).ContextType == "Plan" and "PLAN_VIEW" in target_view:
                offset = ifcopenshell.ifcopenshell_wrapper.float_array_3()
                # A 2mm Z offset to combat Z-fighting in plan or RCPs
                offset[2] = 0.002 if target_view == "PLAN_VIEW" else -0.002
                geom_settings.offset = offset
            geom_settings.set_context_ids(context)
            it = ifcopenshell.geom.iterator(geom_settings, ifc, multiprocessing.cpu_count(), include=drawing_elements)
            processed = set()
            for elem in it:
                processed.add(ifc.by_id(elem.id))
                self.serialiser.write(elem)
                tree.add_element(elem)
            drawing_elements -= processed

    def setup_serialiser(self, ifc, target_view):
        self.svg_settings = ifcopenshell.geom.settings(
            DISABLE_TRIANGULATION=True, STRICT_TOLERANCE=True, INCLUDE_CURVES=True
        )
        self.svg_buffer = ifcopenshell.geom.serializers.buffer()
        self.serialiser = ifcopenshell.geom.serializers.svg(self.svg_buffer, self.svg_settings)
        self.serialiser.setFile(ifc)
        self.serialiser.setWithoutStoreys(True)
        self.serialiser.setPolygonal(True)
        self.serialiser.setUseHlrPoly(True)
        # Objects with more than these edges are rendered as wireframe instead of HLR for optimisation
        self.serialiser.setProfileThreshold(10000)
        self.serialiser.setUseNamespace(True)
        self.serialiser.setAlwaysProject(True)
        self.serialiser.setAutoElevation(False)
        self.serialiser.setAutoSection(False)
        self.serialiser.setPrintSpaceNames(False)
        self.serialiser.setPrintSpaceAreas(False)
        self.serialiser.setDrawDoorArcs(False)
        self.serialiser.setNoCSS(True)
        self.serialiser.setElevationRefGuid(self.camera_element.GlobalId)
        self.serialiser.setScale(1 / 50)
        self.serialiser.setSubtractionSettings(ifcopenshell.ifcopenshell_wrapper.ALWAYS)
        self.serialiser.setUsePrefiltering(True)  # See #3359
        # self.serialiser.setUsePrefiltering(False)  # See #3359
        self.serialiser.setUnifyInputs(True)
        self.serialiser.setSegmentProjection(True)
        if target_view == "REFLECTED_PLAN_VIEW":
            self.serialiser.setMirrorY(True)


Drawer().execute()
```

**Comment by theoryshaw** [comment]
> [@theoryshaw](https://github.com/theoryshaw) for a quick fix right now you can edit the drawing pset to remove any nulls, and then hide any 2d stuff with traits similar to the screenshot.

will do, thanks!

**Comment by aothms** [comment]
> I did encounter another issue which I was unable to dig deeper for `SECTION - SOUTH - GARAGE`:
> 
> ```
> Error: Python: Traceback (most recent call last):
>   File "/home/dion/.config/blender/4.0/scripts/addons/blenderbim/bim/module/drawing/operator.py", line 263, in execute
>     linework_svg = self.generate_linework(context)
>   File "/home/dion/.config/blender/4.0/scripts/addons/blenderbim/bim/module/drawing/operator.py", line 540, in generate_linework
>     self.serialiser.finalize()
>   File "/home/dion/.config/blender/4.0/scripts/addons/blenderbim/libs/site/packages/ifcopenshell/ifcopenshell_wrapper.py", line 1153, in finalize
>     return _ifcopenshell_wrapper.SvgSerializer_finalize(self)
> RuntimeError: An unknown error occurred
> ```

So for me this works. This is likely again due to you using OCCT 7.6. I think we've seen various places where HLR no longer worked. We also still have the "snail shell" issue on revolutions. I think we should collect issues related to 7.6, but not prioritize them at least these few weeks. I would really recommend to setup your local dev env with 7.5. It's just one (slow) git clone + cmake + make install.

**Comment by Moult** [comment]
I can confirm that all issues are now fixed on 7.5. I'll rename this one so we know it's an OCCT >7.5 issue.

**Comment by Moult** [comment]
Closing as the remaining issue is a duplicate of #4924.
